src/asynciterable/zip.ts

```diff
@@ -15,19 +15,25 @@ export class ZipAsyncIterable<TSource> extends AsyncIterableX<TSource[]> {
   // eslint-disable-next-line consistent-return
   async *[Symbol.asyncIterator](signal?: AbortSignal): AsyncIterableIterator<TSource[]> {
     throwIfAborted(signal);
-    const sourcesLength = this._sources.length;
-    const its = this._sources.map((x) => wrapWithAbort(x, signal)[Symbol.asyncIterator]());
-    while (sourcesLength > 0) {
-      const values = new Array(sourcesLength);
-      for (let i = -1; ++i < sourcesLength; ) {
-        const { value, done } = await its[i].next();
-        if (done) {
-          await Promise.all(its.map(returnAsyncIterator));
-          return undefined;
+
+    if (this._sources.length === 0) {
+      return;
+    }
+
+    const iterators = this._sources.map((x) => wrapWithAbort(x, signal)[Symbol.asyncIterator]());
+
+    try {
+      while (1) {
+        const results = await Promise.all(iterators.map((x) => x.next()));
+
+        if (results.some(({ done }) => done)) {
+          return;
         }
-        values[i] = value;
+
+        yield results.map(({ value }) => value);
```

Comment:
The original code constructed a single `values` Array and reused it for each yield, while this constructs a new Array for each `yield.` It was implemented this way to cut down on the number of Array allocations. Is there a compelling reason to use the less efficient approach here?

Reply:
Not intentional, but the new code runs the iterators in parallel with `Promise.all`, which naturally creates an array. I can revert to the serial implementation if you prefer, though I don't see the benefit of doing so.